kubelet/network: report but tolerate errors returned from GetNetNS()

dcbw opened the pull request:
Runtimes should never return "" and nil errors, since network plugin
drivers need to treat netns differently in different cases.  So return
errors when we can't get the netns, and fix up the plugins to do the
right thing.

Namely, we don't need a NetNS on pod network teardown.  We do need
a netns for pod Status checks and for network setup.

@kubernetes/rh-networking @kubernetes/sig-network-bugs @DirectXMan12

dcbw committed:
Revert "Return empty network namespace if the infra container has exited"

This reverts commit fee4c9a7d94e4a44a3b54f04f4e6e17fe2dde3f5.

This is not the correct fix for the problem; and it causes other problems
like continuous:

docker_sandbox.go:234] NetworkPlugin cni failed on the status hook for pod
"someotherdc-1-deploy_default": Unexpected command output nsenter: cannot
open : No such file or directory with error: exit status 1

Because GetNetNS() is returning an empty network namespace.  That is
not helpful nor should really be allowed; that's what the error return
from GetNetNS() is for.

dcbw committed:
kubelet/network: return but tolerate errors returned from GetNetNS()

Runtimes should never return "" and nil errors, since network plugin
drivers need to treat netns differently in different cases.  So return
errors when we can't get the netns, and fix up the plugins to do the
right thing.

Namely, we don't need a NetNS on pod network teardown.  We do need
a netns for pod Status checks and for network setup.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43879)
<!-- Reviewable:end -->

dcbw commented:
Need to figure out the CI failures; in my queue to do.

dcbw committed:
dockershim: don't spam logs with pod IP errors before networking is ready

GenericPLEG's 1s relist() loop races against pod network setup.  It
may be called after the infra container has started but before
network setup is done, since PLEG and the runtime's SyncPod() run
in different goroutines.

Track network setup status and don't bother trying to read the pod's
IP address if networking is not yet ready.

See also: https://bugzilla.redhat.com/show_bug.cgi?id=1434950

Mar 22 12:18:17 ip-172-31-43-89 atomic-openshift-node: E0322
   12:18:17.651013   25624 docker_manager.go:378] NetworkPlugin
   cni failed on the status hook for pod 'pausepods22' - Unexpected
   command output Device "eth0" does not exist.

danwinship reviewed (approved):
LGTM

freehan commented:
/approve
just need to fix the test.

yujuhong commented:
Please leave out the changes to the `dockertools/` package. I plan to remove the majority of the code in that package in a couple of weeks (see #43884).

dcbw commented:
@yujuhong @danwinship @freehan rebased and changed a bit, PTAL thanks!

WIZARD-CXY commented:
I wonder when will this pr will be merged to fix #44307

yujuhong commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -281,11 +288,19 @@ func (ds *dockerService) getIP(sandbox *dockertypes.ContainerJSON) (string, erro
-	} else if IP != "" {
-		return IP, nil
+
+	// Don't bother getting IP if networking isn't ready
```
If kubelet restarts, it'd lose the in-memory cache and not get the IP for any sandbox.
It's hard to eliminate the race solely based on the sandbox ID since we don't get it until the underlying container is created. An alternative is to use the sandbox name as a key in the map.
 - Add the key to the map *before* creating the container.
 - If the key exists, assume the network is being set up, and don't check the IP.
 - If the key does not exist, get the IP as usual. 
 - Delete the key once the setup is done.

dcbw replied:
@yujuhong good point; though it's hard to use namespace/name because we can't easily get that from RemovePodSandbox() which is where we actually need to delete the item from the networkReady map.  We need to delete there because the testcases rightfully point out that after StopPodSandbox() a status request shouldn't return the Pod IP, so if we delete from the map in StopPodSandbox() then "if key does not exist, get the IP as usual" breaks that testcase.

And you're right, we don't get the sandbox ID until the container is created.  The real race is between ListPodSandbox() and RunPodSandbox(), where there is a "large" amount of time between creating the docker container and finishing the network setup, while in between those operations PLEG has access to the sandbox ID via ListPodSandbox().

So I suggest a hybrid approach.  We still use the pod sandbox ID as the key, but we add to the networkReady map with 'false' right after creation in RunPodSandbox(), then switch to 'true' when network has completed.  We flip back to 'false' in StopPodSandbox(), and delete from the map in RemovePodSandbox().

Then we use your suggestion for GetPodSandboxStatus() where if the key is not present, we assume network ready.  That should handle kubelet restart by allowing the status to proceed, but block PLEG for pods not-yet-setup.

Seem OK?

yujuhong replied:
Sounds good.

For the long run, we should probably also checkpoint the actual name we've built for the container (which should always be unique), so that we can always look it up.
/cc @freehan

dcbw commented:
@k8s-bot cvm gce e2e test this

dcbw commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

yujuhong commented:
LGTM. @freehan, do you want to take a look? 

/approve

freehan commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -82,6 +102,8 @@ func (ds *dockerService) RunPodSandbox(config *runtimeapi.PodSandboxConfig) (str
 		return "", fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
 	}
 
+	ds.setNetworkReady(createResp.ID, false)
```
I cannot think of a better way to address this. Not much options:

1. kubelet needs to know the RunPodSandbox has not finish. Do not call PodSandboxStatus (Not possible since sandbox ID is unknown to kubelet)
2. dockershim needs to know RunPodSandbox has not finish. Skip getting ip from network plugin. 

I am willing to settle with the current solution.

dcbw replied:
Yes, still a small race.  But IMHO SetUpPod is what takes the longest time, and that's what was causing the race.  We can't really address that unless we can use name/namespace as the networkReady key, and we can't do that unless we can add that to the checkpoint.

One option I toyed with was adding docker container labels with the name/namespace, which we could get from ListContainers().  What do you think?  We already use labels int he Checkpoint code, and the rkt runtime does something similar.  However, I'd like to do that as a follow-up PR if that seems workable...

freehan reviewed (commented):
just nits

dcbw commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -281,11 +309,18 @@ func (ds *dockerService) getIP(sandbox *dockertypes.ContainerJSON) (string, erro
 	}
+
+	ip, err := ds.getIPFromPlugin(sandbox)
+	if err == nil {
```
@freehan no, that's mostly the point of this PR.  We don't want to warn in *this* function, but we want to pass the warning up to higher levels, eg the caller of PodSandboxStatus() so it can do whatever it wants with that information, whether it's warn or ignore.

dcbw commented:
@freehan since your LGTM doesn't seem to be enough, who else needs to lgtm it?

knobunc reviewed (approved):
LGTM

yujuhong commented:
/lgtm

yujuhong commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this
